Xamarin.Forms.Core/DataTemplateSelector.cs

```diff
@@ -1,15 +1,39 @@
 using System;
+using System.Collections.Generic;
 
 namespace Xamarin.Forms
 {
 	public abstract class DataTemplateSelector : DataTemplate
 	{
+		Dictionary<Type, DataTemplate> _dataTemplates = new Dictionary<Type, DataTemplate>();
+
 		public DataTemplate SelectTemplate(object item, BindableObject container)
 		{
-			DataTemplate result = OnSelectTemplate(item, container);
-			if (result is DataTemplateSelector)
-				throw new NotSupportedException("DataTemplateSelector.OnSelectTemplate must not return another DataTemplateSelector");
-			return result;
+			var listView = container as ListView;
+
+			var recycle = listView == null ? false :
+				(listView.CachingStrategy & ListViewCachingStrategy.RecycleElementAndDataTemplate) ==
+					ListViewCachingStrategy.RecycleElementAndDataTemplate;
+
+			DataTemplate dataTemplate = null;
+			if (recycle && _dataTemplates.TryGetValue(item.GetType(), out dataTemplate))
+				return dataTemplate;
+
+			dataTemplate = OnSelectTemplate(item, container);
+			if (dataTemplate is DataTemplateSelector)
+				throw new NotSupportedException(
+					"DataTemplateSelector.OnSelectTemplate must not return another DataTemplateSelector");
+
+			if (recycle)
+			{
+				if (!dataTemplate.CanRecycle)
+					throw new NotSupportedException(
+						"RecycleElementAndDataTemplate requires DataTemplate activated with ctor taking a type.");
```

Comment:
If we don't add the new caching strategy `RecycleElementAndDataTemplate` then we cannot raise this exception.